sklearn/utils/tests/test_validation.py

```diff
@@ -2124,3 +2125,67 @@ def __init__(self):
             self.schema = ["a", "b"]
 
     assert not _is_polars_df(LooksLikePolars())
+
+
+def test_check_array_writeable_np():
+    """Check the behavior of check_array when a writeable array is requested
+    without copy if possible, on numpy arrays.
+    """
+    X = np.random.uniform(size=(10, 10))
+
+    out = check_array(X, copy=False, force_writeable=True)
+    # X is already writeable, no copy is needed
+    assert np.may_share_memory(out, X)
+    assert out.flags.writeable
+
+    X.flags.writeable = False
+
+    out = check_array(X, copy=False, force_writeable=True)
+    # X is not writeable, a copy is made
+    assert not np.may_share_memory(out, X)
+    assert out.flags.writeable
+
+
+def test_check_array_writeable_mmap():
+    """Check the behavior of check_array when a writeable array is requested
+    without copy if possible, on a memory-map.
+
+    A common situation is when a meta-estimators run in parallel using multiprocessing
+    with joblib, which creates read-only memory-maps of large arrays.
+    """
+    X = np.random.uniform(size=(10, 10))
+
+    mmap = create_memmap_backed_data(X, mmap_mode="w+")
+    out = check_array(mmap, copy=False, force_writeable=True)
+    # mmap is already writeable, no copy is needed
+    assert np.may_share_memory(out, mmap)
+    assert out.flags.writeable
+
+    mmap = create_memmap_backed_data(X, mmap_mode="r")
+    out = check_array(mmap, copy=False, force_writeable=True)
+    # mmap is read-only, a copy is made
+    assert not np.may_share_memory(out, mmap)
+    assert out.flags.writeable
+
+
+def test_check_array_writeable_df():
+    """Check the behavior of check_array when a writeable array is requested
+    without copy if possible, on a dataframe.
+    """
+    pd = pytest.importorskip("pandas")
+
+    X = np.random.uniform(size=(10, 10))
+    df = pd.DataFrame(X, copy=False)
+
+    out = check_array(df, copy=False, force_writeable=True)
+    # df is backed by a writeable array, no copy is needed
+    assert np.may_share_memory(out, df)
+    assert out.flags.writeable
+
+    X.flags.writeable = False
+    df = pd.DataFrame(X, copy=False)
+
+    out = check_array(df, copy=False, force_writeable=True)
+    # df is backed by a read-only array, a copy is made
+    assert not np.may_share_memory(out, df)
+    assert out.flags.writeable
```

Comment:
I think we need a similar test for array API inputs using the generic `yield_namespace_device_dtype_combinations` helper.

The array API states the following: https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html

So maybe we could just raise an exception when `copy is False and force_writeable and not _is_numpy_namespace(xp)`.

Reply:
If we decide that we should not raise an exception in that case for some reason (e.g. by always triggering a copy for safety?), then we should have a dedicated test such as:

```python
xp = pytest.importorskip("array_api_strict")
X_np = np.random.uniform(size=(10, 10))
X_np.flags.writeable = False
X_np_copy = X_np.copy()

X_xp = xp.asarray(X_np)
with sklearn.config_context(array_api_dispatch=True):
    X_xp_checked = check_array(X_xp, copy=False, force_writeable=True)
    out_ns, is_array_api = get_namespace(X_xp_checked)
    assert is_array_api
    assert out_ns == xp
    assert_allclose(_convert_to_numpy(X_xp_checked), X_np)
    X_xp_checked[:] = 0

assert_allclose(X_np, X_np_copy)
```

And maybe something similar for PyTorch. I am not sure if it's possible to create readonly PyTorch tensors. On CPU it might be possible with memmaping?

However I am not sure this is what we want...

Reply:
The array API draft spec for the 2024 version was updated to mention readonly flags exposed by the DLPack interchange protocol:

https://github.com/data-apis/array-api/pull/749/files

However numpy 1.16.6 does not support this (yet) and raises instead...

```
In [1]: import numpy as np

In [2]: a = np.random.randn(10)

In [3]: a.flags.writeable = False

In [4]: a.__dlpack__()
---------------------------------------------------------------------------
BufferError                               Traceback (most recent call last)
Cell In[4], line 1
----> 1 a.__dlpack__()

BufferError: Cannot export readonly array since signalling readonly is unsupported by DLPack.
```

maybe we can reconsider inspecting `__dlpack__` attributes later, once it's more widely adopted by libraries.

Reply:
In retrospect, I think I would be in favor of raising an exception as first suggested in https://github.com/scikit-learn/scikit-learn/pull/29018/files#r1644733447.

However to keep that PR minimally focused on the changes actually needed to fix the blocking bugs for 1.5.1, we can defer the new array API specific tests and the exception to a dedicated follow-up PR for 1.6 and only fix:

https://github.com/scikit-learn/scikit-learn/pull/29018/files#r1645720337

as part of the current PR.

Reply:
Alright, let's do that in a follow-up PR, the current fix should be enough for 1.5.1